Make Config generator test only for down-casts

ericstj opened the pull request:
Fix #94547 

We were hitting compiler errors when the generator emitted test casts for value types.  Since those can never be true (value types cannot be derived from, and the compiler can see if the cast will succeed or not).

Fix this by only doing a test cast when we are trying to do a runtime downcast.

ericstj committed:
Make Config generator test for downcasts

We were hitting compiler errors when the generator emitted test casts
for value types.  Since those can never be true (value types cannot be
derived from, and the compiler can see if the cast will succeed or not).

Fix this by only doing a test cast when we are trying to do a runtime
downcast.

eerhardt commented on src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Specs/Types/CollectionSpec.cs:
```diff
@@ -23,6 +23,8 @@ internal abstract record CollectionWithCtorInitSpec : CollectionSpec
 
         public required CollectionPopulationCastType PopulationCastType { get; init; }
+
+        public required bool ShouldTestCast {get; init; }
```
(nit)
```suggestion
        public required bool ShouldTestCast { get; init; }
```

eerhardt replied:
Would it be better to name this `ShouldTryCast`? Or invert it to `UseDirectCast`?

I'm not sure I understood what "test cast" meant right away.

ericstj replied:
Yeah, the naming is hard.  I think `ShouldTryCast` is better.

ericstj replied:
FWIW it is called `test` in the C# language https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/type-testing-and-cast

eerhardt replied:
Yeah - it might just be me then. I guess I hadn't heard that phrase used - but I'm not really a language expert.

If you want to name it back, I wouldn't object.

tarekgh commented on src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs:
```diff
@@ -431,13 +435,15 @@ private TypeSpec CreateDictionarySpec(TypeParseInfo typeParseInfo, ITypeSymbol k
                 TypeRef keyTypeRef = EnqueueTransitiveType(typeParseInfo, keyTypeSymbol, DiagnosticDescriptors.DictionaryKeyNotSupported);
                 TypeRef elementTypeRef = EnqueueTransitiveType(typeParseInfo, elementTypeSymbol, DiagnosticDescriptors.ElementTypeNotSupported);
 
+                Debug.Assert(!shouldTestCast || !type.IsValueType, "Should not test cast for value types.");
```
>Debug.Assert(!shouldTestCast || !type.IsValueType, "Should not test cast for value types."); [](http://example.com/codeflow?start=16&length=92)

what happen when having a struct implementing `IReadonlyDictionary`?

ericstj replied:
The `IReadonlyDictionary` behavior is only for when the actual type is `IReadonlyDictionary`, not for a type implementing `IReadonlyDictionary`.  This is what I tried to note in the comments I added.

tarekgh commented on src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs:
```diff
@@ -514,12 +528,14 @@ private TypeSpec CreateEnumerableSpec(TypeParseInfo typeParseInfo)
 
                 TypeRef elementTypeRef = EnqueueTransitiveType(typeParseInfo, elementType, DiagnosticDescriptors.ElementTypeNotSupported);
 
+                Debug.Assert(!shouldTestCast || !type.IsValueType, "Should not test cast for value types.");
```
> Debug.Assert(!shouldTestCast || !type.IsValueType, "Should not test cast for value types."); [](http://example.com/codeflow?start=15&length=93)

similar to previous question

ericstj replied:
Same answer - the test down cast is only applicable when the bound type is itself `IReadonlyDictionary`, not for types that implement `IReadonlyDictionary`.

eerhardt reviewed (commented):
This looks good. Thanks for getting this fixed.

I just had minor suggestions.

tarekgh commented on src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.TestClasses.Collections.cs:
```diff
@@ -388,5 +389,78 @@ public class Element
+            bool ICollection<string>.Remove(string item) => _collection.Remove(item);
+        }
+
+        public struct DictionaryStructExplicit : IDictionary<string, string>
```
> public struct DictionaryStructExplicit : IDictionary<string, string> [](http://example.com/codeflow?start=7&length=69)

Can we add test for struct implmenting IReadOnlyDictionary and IReadOnlyCollection?

ericstj replied:
I can - both will be "unsupported" since they don't implement the writable interfaces.  The downcast check only happens when the bound type itself is IRO interface - in which case it's always safe to test cast the interface object.

ericstj replied:
I added those tests to demonstrate that they do nothing.

tarekgh replied:
LGTM

tarekgh reviewed (approved):
Added a couple of questions. LGTM.

ericstj committed:
Rename ShouldTestCast -> ShouldTryCast

ericstj committed:
Add tests for readonly collection structs

These are unsupported and ignored by the generator and a diagnostic is
emitted.

Binding data to them does nothing.

eiriktsarpalis reviewed (approved):
Looks good, although it seems like the flag is only kicking in when predetermined pairs of types are detected. Can the source generator produce casts for types other than these collection types, and if so couldn't the `shouldTryCast` value be determined in the general case? (i.e. by looking at the subtype relationship between the source and target types)

ericstj commented:
> Looks good, although it seems like the flag is only kicking in when predetermined pairs of types are detected

It's everywhere that `PopulationCastType` is specified and used.

> Can the source generator produce casts for types other than these collection types

I audited this file for uses of `is` in generated code and didn't see anything else that stood out as applying to the target type.  Most checks are around the Configuration passed in at runtime.  A couple cases of `null` but those were guarded with `CanBeNull` checks.  It was a good suggestion to look for variants, and I couldn't find any.

steveharter commented on src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Baselines/netcoreapp/Version0/Collections.generated.txt:
```diff
@@ -36,12 +36,12 @@ namespace Microsoft.Extensions.Configuration.Binder.SourceGeneration
         #region IConfiguration extensions.
         /// <summary>Attempts to bind the configuration instance to a new instance of type T.</summary>
-        [InterceptsLocation(@"src-0.cs", 12, 17)]
+        [InterceptsLocation(@"src-0.cs", 13, 17)]
```
Good to see this change in the "version 0" baseline - did you encounter any test issues or did you just remember to do this?
